[Feat] Baseline support (2/2)

yashnator opened the pull request:
This pull request is part (1/2) of the baseline support for RootIO. 

The PR provides functions to:
- Create a new TTree using keyword arguments
- Adds docstrings and documenter.jl
- Adds unit tests
- Adds support for writing boolean and strings(as char*)

Example:
```julia
import RootIO, ROOT
using DataFrames
file = ROOT.TFile!Open("example.root", "RECREATE")

name = "example_tree"
title = "Example TTree"
data = (col_float=rand(Float64, 3), col_int=rand(Int32, 3))
tree = RootIO.TTree(file, name, title; data...)

RootIO.Write(tree)
ROOT.Close(file)
```

Checklist before requesting a review:
- [x] Tested locally

yashnator committed:
[Feat] Add support for keyword-argument constructor

yashnator committed:
[Chore] Add doc strings to the functions

yashnator committed:
[Test] Add tests for primitive integer and floats

Moelf commented on Project.toml:
```diff
@@ -7,3 +7,4 @@ version = "0.1.0"
 CxxWrap = "1f15a43c-97ca-5a2a-ae31-89f07a497df4"
 ROOT = "1706fdcc-8426-44f1-a283-5be479e9517c"
 Tables = "bd369af6-aec1-5ad0-b16a-f7cc5008161c"
+Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
```
This should only be in the extra section for tests

yashnator replied:
I missed it, my bad. Resolved!

yashnator committed:
[Feat] Add support for strings and booleans

- Adds support for writing strings and booleans
- Adds tests for writing strings and booleans

grasph commented on src/RootIO.jl:
```diff
@@ -2,21 +2,106 @@ module RootIO
+A struct representing a ROOT TTree with its associated branches and file.
+
+# Fields
+- `_ROOT_ttree`: The ROOT TTree object.
```
You don't want the private field to appear in the generated documentation.
 ⇒ put the field documentation as comments, so it is kept in the code, while being skip by Documenter.

grasph commented on src/RootIO.jl:
```diff
@@ -2,21 +2,106 @@ module RootIO
+# Returns
+- A RootIO `TTree` object containing the ROOT TTree and its branches.
+"""
+function _makeTTree(file::CxxWrap.CxxWrapCore.CxxPtr{ROOT.TFile}, name::String, title::String, _branch_types, _branch_names)
```
no need for underscores for the arguments:
-  _branch_types → branch_types
- _ branch_names → branch_names

grasph commented on src/RootIO.jl:
```diff
@@ -2,21 +2,106 @@ module RootIO
+- A RootIO `TTree` object containing the ROOT TTree and its branches.
+"""
+function _makeTTree(file::CxxWrap.CxxWrapCore.CxxPtr{ROOT.TFile}, name::String, title::String, _branch_types, _branch_names)
     _tree = ROOT.TTree(name, title)
```
remove the underscores from all the local variables

grasph commented on src/RootIO.jl:
```diff
@@ -2,21 +2,106 @@ module RootIO
+"""
+    TTree(file::CxxWrap.CxxWrapCore.CxxPtr{ROOT.TFile}, name::String, title::String, data)
+
+Creates a new ROOT TTree and fills it with the provided data.
```
According to the specs. the method should only create the tree.

grasph commented on src/RootIO.jl:
`````diff
@@ -2,21 +2,106 @@ module RootIO
+````
+"""
+function TTree(file::CxxWrap.CxxWrapCore.CxxPtr{ROOT.TFile}, name::String, title::String, data)
+    println("Creating a new TTree")
`````
to remove

grasph commented on src/RootIO.jl:
```diff
@@ -49,6 +122,80 @@ function TTree(file::CxxWrap.CxxWrapCore.CxxPtr{ROOT.TFile}, name::String, title
+"""
+    TTree(file::CxxWrap.CxxWrapCore.CxxPtr{ROOT.TFile}, name::String, title::String; kwargs...)
+
+Creates a new ROOT TTree and fills it with the provided data.
```
according to specs the method should only create the TTree

grasph commented on src/RootIO.jl:
```diff
@@ -49,6 +122,80 @@ function TTree(file::CxxWrap.CxxWrapCore.CxxPtr{ROOT.TFile}, name::String, title
+"""
 function Fill(tree::TTree, data)
     if Tables.istable(data)
         println("Filling the TTree with a table")
```
to remove or change in a `@debug`

grasph commented on src/RootIO.jl:
```diff
@@ -57,10 +204,23 @@ function Fill(tree::TTree, data)
         end
     else
         println("Filling the TTtree with a row")
```
to remove

grasph commented on src/RootIO.jl:
```diff
@@ -57,10 +204,23 @@ function Fill(tree::TTree, data)
         for i in 1:length(tree._branch_array)
             if typeof(row[i]) <: CxxWrap.StdVector
                 ROOT.SetObject(tree._branch_array[i], row[i])
+            elseif typeof(row[i]) == String
```
use `isa` function

grasph commented on src/RootIO.jl:
```diff
@@ -57,10 +204,23 @@ function Fill(tree::TTree, data)
+                GC.@preserve str begin
+                    ROOT.SetBranchAddress(tree._ROOT_ttree, string(tree._branch_names[i]), convert(Ptr{Int8}, pointer(str)))
+                end
+            elseif typeof(row[i]) == Bool
```
use `isa()`

grasph commented on src/RootIO.jl:
```diff
@@ -57,10 +204,23 @@ function Fill(tree::TTree, data)
+                    ROOT.SetBranchAddress(tree._ROOT_ttree, string(tree._branch_names[i]), convert(Ptr{Int8}, pointer(b)))
+                end
             else
                 ROOT.SetAddress(tree._branch_array[i], Ref(row[i]))
```
You need to keep an instance of Ref(row[i]) to then preserve it from GC.

You have an issue in L229 below. What needs to be preserved is the data variables, because they are used by the backend code of `Fill()`, while they are not referenced in the arguments. See the `read_tree3.jl` example of `ROOT.jl`

yashnator replied:
I read more about Julia GC. I've preserved the entire row and kwargs(in the above function). Please look into it if it's correct now

grasph commented:
Please find my comments below,

https://github.com/JuliaHEP/RootIO.jl/blob/49a741dc10d9c17fd20c2f54e1f8501a4865fa06/src/RootIO.jl:

LL 129ff: ~I don't get how you specify to ROOT the branches that hold array lengths. Do you have a test that shows that it works ?~ Remove code from the c-array support as the implementation is not completed and it is the topic of PR #3.

LL 50 and 59 bufsize parameter to change from 100 to 3200, the native function default.

L 144: `eletype` -> `eltype` ?

yashnator committed:
[Fix] Refactor code and changing convention

- Removes redundant code
- Refactors and renames variables
- Only types inferred from constructor, data is
not written
- Updated tests to follow convention
